Fix overlapping crossbars

pedroilidio opened the pull request:
Three main points are relevant in this PR:

1. It fixes #86.

2. I also modified the `sign_array` function, since I was again running into [#82](https://github.com/maximtrp/scikit-posthocs/issues/82). The modification dispenses the epsilon added by #81 and discussed in [#82](https://github.com/maximtrp/scikit-posthocs/issues/82). Please let me know if this is incorrect. I made sure the copy problem pointed out by [#77](https://github.com/maximtrp/scikit-posthocs/issues/77) remains solved.

3. To ensure these problems do not come up in the future I added unit tests based on [#86](https://github.com/maximtrp/scikit-posthocs/issues/86) and [#82](https://github.com/maximtrp/scikit-posthocs/issues/82).

pedroilidio committed:
fix overlapping crossbars

pedroilidio committed:
remove unecessary line

pedroilidio committed:
remove unecessary lines

pedroilidio committed:
minor fix

maximtrp commented:
Thank you! I'm gonna review all the related code in your pull request and this package

pedroilidio commented:
Indeed, @maximtrp, I'll add as many tests for the CDDs as I can :)

pedroilidio commented:
I apologize, I was quite wrong [here](https://github.com/maximtrp/scikit-posthocs/pull/87#issuecomment-2806969528):

> Furthermore, I replaced the iterative algorithm to find the crossbar levels:
> [...]


The maximal cliques approach does not ensure different crossbars at each level. This is more complicated: it's a graph coloring problem in the complement graph of the crossbar graph (`on_same_level`) (see [this SE answer](https://cs.stackexchange.com/a/42166)).

I'll revert back to the greedy approach we were using. Another (maybe overkill) option: I could adapt the implementation of [NetworkX's approach](https://networkx.org/documentation/stable/_modules/networkx/algorithms/coloring/equitable_coloring.html#equitable_color) in a separate module, avoiding to add NetworkX as a new dependency. I could do the same for the maximal cliques function I needed to implement for the CDDs.

pedroilidio committed:
Fix point order in each crossbar (#89)

Pointed out here:
https://github.com/maximtrp/scikit-posthocs/pull/87#issuecomment-2812740771

And also in #89.

pedroilidio commented:
It looks better now, but it suspiciously changed from my previous figure (pairs of points are gone). I will write several tests in a further commit.

![fixed](https://github.com/user-attachments/assets/aba30bcf-3471-4c77-8f42-247eb26a8b97)


```python
import numpy as np
import matplotlib.pyplot as plt
from pandas import DataFrame, Series

from scikit_posthocs import critical_difference_diagram

n = 10
rng = np.random.default_rng(0)
rank_values = rng.uniform(size=n)
index = np.arange(n).astype(str)
ranks = Series(rank_values, index=index)

triu = np.triu_indices(n, k=1)
values = rng.integers(0, 2, size=len(triu[0]), dtype=np.int8)
sig_array = np.ones((n, n), dtype=np.int8)
sig_array[triu] = sig_array.T[triu] = values

sig_matrix = DataFrame(sig_array, index=index, columns=index)

plt.figure(figsize=(5, 3))
output = critical_difference_diagram(
    ranks, sig_matrix, crossbar_props={"marker": "s", "linestyle": ":", "alpha": .3}
)
plt.tight_layout()
plt.savefig("fixed.png")
```

pedroilidio committed:
minor refactoring

pedroilidio committed:
comment tweak

pedroilidio committed:
remove unnecessary line

maximtrp commented:
Thank you once again! Now not all tests are passed ☝️

pedroilidio commented:
Thanks for following this thread, @maximtrp! I'll be busier for some days, so I'll get back to working on this at the end of the next week. For now I'll convert the PR to a draft PR until it's ready for merging. Next steps are:

- [ ] Solve failing tests.
- [ ] Add way more unit tests, e.g.:
  - Assert ranks are ordered in each crossbar (this fixes #89).
  - Assert crossbars do not ever overlap.
  - Assert maximal cliques function is properly working (maybe compare several random cases with NetworkX results)
  - etc.
- [ ] Decide: for the greedy placement of crossbars in the levels, should we prioritize wider bars, left-most bars, or maybe a mixture of both? (I'm talking specifically about the following line of code)
https://github.com/maximtrp/scikit-posthocs/blob/a8d50cb03eda1f97c40e823ac84f30b097933dc8/scikit_posthocs/_plotting.py#L531

maximtrp commented:
I hope I've fixed it in https://github.com/maximtrp/scikit-posthocs/commit/e8f6195f10804b1ad74b3a34e883ebd0792db39d Please open an issue if anything else is wrong.
